base/iterators.jl

```diff
@@ -988,7 +988,13 @@ cycle(xs, n::Integer) = flatten(repeated(xs, n))
 
 eltype(::Type{Cycle{I}}) where {I} = eltype(I)
 IteratorEltype(::Type{Cycle{I}}) where {I} = IteratorEltype(I)
-IteratorSize(::Type{Cycle{I}}) where {I} = IsInfinite() # XXX: this is false if iterator ever becomes empty
+function IteratorSize(::Type{Cycle{I}}) where I
+    # TODO: find a better way of communicating the size of a cycle
+    # IsInfinite() would be false if iterator ever becomes empty
+    IteratorSize(I) === IsInfinite() ? IsInfinite() : SizeUnknown()
+end
+IteratorSize(it::Cycle) = isempty(it.xs) ? HasLength() : IsInfinite()
```

Comment:
Do we have other examples where `IteratorSize(it)` and `IteratorSize(typeof(it))` (potentially) disagree? Makes me a bit uncomfortable.

And if `it.xs` is stateful, it seems that `IsInfinite` might turn out to be wrong, because `isempty(it.xs)` might become true.

Reply:
> Do we have other examples where IteratorSize(it) and IteratorSize(typeof(it)) (potentially) disagree? Makes me a bit uncomfortable.

No, this is the first case, and changes the docs to suggest the possibility.

> And if it.xs is stateful, it seems that IsInfinite might turn out to be wrong, because isempty(it.xs) might become true.

IMO that's because Iterators.cycle on stateful iterators is broken. It is explicitly documented to be infinite for non-empty arguments.